SOC-735 Implement Forgot Password request email.

garthwebb opened the pull request:
https://wikia-inc.atlassian.net/browse/SOC-735

Commit:
SOC-735 Ran stylize.php on all UserLogin files

Commit:
Merge branch 'email-sprint-15' into SOC-735

Commit:
SOC-735 Implement forgot password email

Commit:
Merge branch 'email-sprint-15' into SOC-735

garthwebb commented on extensions/wikia/Email/Controller/ForgotPasswordController.class.php:
```diff
@@ -14,75 +14,52 @@
  * @package Email\Controller
```
Hello, and welcome to another installment of "Review the long email change set."  I'll be your host, @garthwebb, and today we'll be looking at a real classic.

garthwebb commented on extensions/wikia/Email/Controller/ForgotPasswordController.class.php:
```diff
@@ -14,75 +14,52 @@
 class ForgotPasswordController extends EmailController {
+	private $tempPass;
 
 	public function initEmail() {
```
You should probably just view the full file for this one, since it really should appear as a new file anyway.  Since this was the first example controller, there were tons of things that changed.

garthwebb commented on extensions/wikia/Email/Email.i18n.php:
```diff
@@ -9,14 +9,6 @@
 	'emailext-anonymous-editor' => 'A Wikia fan',
 	'emailext-no-reply-name' => 'No Reply',
 	'emailext-fans-tagline' => 'The Social Universe for Fans by Fans',
```
Just ripped these out since they aren't the right split of the messages and are the right copy and are in the wrong i18n file.

Commit:
SOC-735 Remove unnecessary 'use' line.

garthwebb commented on extensions/wikia/Email/EmailController.class.php:
```diff
@@ -680,4 +665,44 @@ protected function getMessage() {
+		] );
+	}
+
+	private function trackEmailError( \Exception $e ) {
```
I'm being extra careful with how I get these values here since the error thrown could have been because some of these required values may have been empty.

Commit:
SOC-735 Added some docs

garthwebb commented on includes/wikia/services/UserService.class.php:
```diff
@@ -43,6 +43,36 @@ public function getUsers( $ids ) {
+
 	/** Helper methods for getUsers */
 
 	/**
```
And that's the show.  Please tip your server!

Commit:
SOC-735 Don't forget tracking

jsutterfield commented on extensions/wikia/Email/i18n/ForgotPassword.i18n.php:
```diff
@@ -0,0 +1,20 @@
+	'emailext-password-summary' => 'It looks like you’re having trouble logging in. Sorry about that! As much as we’d like to just send you your password, for security reasons, we’ll need you to login with a temporary password.',
+	'emailext-password-intro' => 'Please see your temporary password below:',
+	'emailext-password-unrequested' => "If you didn't request a new password, don't worry! Your account is safe and secure. You can ignore this email and log in to Wikia with your old password.",
+	'emailext-password-questions' => 'Questions or concerns? Feel free to [http://community.wikia.com/wiki/Special:Contact/account-issue contact us].',
```
Is there some this url can be generated programmatically based on the target user's language settings? Seems like we should be directing user's to their languages community.wikia.com if it's available.

jsutterfield commented on extensions/wikia/Email/i18n/ForgotPassword.i18n.php:
```diff
@@ -0,0 +1,20 @@
+	'emailext-password-summary' => 'Summary text for the forgot password email',
+	'emailext-password-intro' => 'Text shown just above the new temporary password',
+	'emailext-password-unrequested' => "Instructions for what to do if you didn't request a new password",
+	'emailext-password-questions' => 'Text on how to contact support.  The link here should be updated to point at the Wikia contact page for the appropriate language',
```
Oh, I see you're calling out that i18n'd community URL here. If there was a way to have the code construct the URL I think that would be preferred, but if not this works fine too.

garthwebb replied:
Not as far as I know.  Sometimes you can put a language prefix in front of the English hostname, sometimes it lives at a different URL altogether.

jsutterfield commented on extensions/wikia/UserLogin/UserLoginSpecialController.class.php:
```diff
@@ -589,37 +591,54 @@ public function mailPassword() {
+		] );
+	}
+
+	private function setErrorResponse( $key ) {
```
Is there a reason to have the `$key` argument here as part of the method signature? It's not used anywhere in the method since we're calling `func_get_args`

kvas-damian replied:
Maybe we could be explicit about input parameters and use variadic function here: http://php.net/manual/en/migration56.new-features.php#migration56.new-features.variadics ?

kvas-damian commented on extensions/wikia/Email/i18n/ForgotPassword.i18n.php:
```diff
@@ -0,0 +1,20 @@
+	'emailext-password-unrequested' => "Instructions for what to do if you didn't request a new password",
+	'emailext-password-questions' => 'Text on how to contact support.  The link here should be updated to point at the Wikia contact page for the appropriate language',
+	'emailext-password-signature' => 'The signature for this email',
+);
```
Missing new line at the end of file ;)

garthwebb replied:
Should be there now

kvas-damian commented:
Please fix unit tests

kvas-damian commented on extensions/wikia/Email/Controller/ForgotPasswordController.class.php:
```diff
@@ -15,74 +14,54 @@
+			$this->getMessage( 'emailext-password-footer-2' )->escaped(),
+		];
 	}
 }
```
Don't we need ```getEmailSpecificFormFields``` in this class?

garthwebb replied:
Nope.  The parent defines everything we need, which is basically just the targetUser and language.  The temp pass will be generated if not passed.

kvas-damian replied:
I'm not sure generating new temp password on every test email send is best idea ;)

garthwebb replied:
What problems would it cause?  It will only be used internally by dev, QA and LOC.

kvas-damian replied:
Agree, it should be ok.

Commit:
SOC-735 Fixed unit tests.

Also removed one set of test parameters.  They basically tested that when the LoginForm object returned a fatal status object, that it returned a fatal status object :/

Given that we don't even use the LoginForm object to send the email anymore, this doesns't make sense to keep

Commit:
SOC-735 Responded to code review

* newline
* new variadic function stuffs

garthwebb commented on extensions/wikia/UserLogin/tests/UserLoginTest.php:
```diff
@@ -143,8 +143,6 @@ public function loginDataProvider() {
 			'loginToken' => '',
```
Several spacing changes and ```array()``` to ```[]``` changes here.

garthwebb commented on extensions/wikia/UserLogin/tests/UserLoginTest.php:
```diff
@@ -313,114 +315,112 @@ public function testMailPassword( $requestParams, $mockWgUserParams, $mockAuthPa
+		];
 		$expMsg6 = wfMessage( 'userlogin-error-throttled-mailpassword', round( F::app()->wg->PasswordReminderResendTime, 3 ) )->escaped();
 
 		// User - mail error
```
Removed the test #7 params since that test was no longer relevant.

garthwebb commented on extensions/wikia/UserLogin/tests/UserLoginTest.php:
```diff
@@ -313,114 +315,112 @@ public function testMailPassword( $requestParams, $mockWgUserParams, $mockAuthPa
+			]
+		];
 
 		// User - email sent
```
Removing mocking of mailPasswordInternal so getting rid of these params.

Commit:
SOC-735 Removed commented out lines

garthwebb commented on extensions/wikia/UserLogin/tests/UserLoginTest.php:
```diff
@@ -313,114 +315,110 @@ public function testMailPassword( $requestParams, $mockWgUserParams, $mockAuthPa
-				)
-			)
-		);
+			'params' => [
```
The DB for user ```WikiaUser``` doesn't have this ID anymore and I don't know why both have to be specified.  Removed it and added mEmail which the new controller checks for.

kvas-damian commented on extensions/wikia/UserLogin/UserLoginSpecialController.class.php:
```diff
@@ -589,39 +591,75 @@ public function mailPassword() {
 		}
 	}
 
+	private function setErrorResponse( ...$params ) {
```
While ```$key``` is required, I'd suggest putting it as a separate params in function signature. What do you think?

garthwebb replied:
It would kind of complicate things.  If I specified ```$key``` here, it would no longer be in $params.  At some point I would have to add it back to $params without ever being used directly.

garthwebb replied:
Just added some PHP docs to make it more clear what these methods expect.

kvas-damian replied:
Why you would need to add it back to params?

In ```setResponseGeneric()``` you could try:
````
$msg = wfMessage( $key )->pararms( $params )->$postProcess();
````
Or many other combination message parameters including extracting array by ```...$params```. ```wfMessage``` is pretty flexible about that.

garthwebb replied:
Oh, I didn't realize wfMessage could take an array as a second arg, and the unpacking of the array via ...$params make sense too.

kvas-damian replied:
We're learning all the time ;)

kvas-damian commented on extensions/wikia/Email/Controller/ForgotPasswordController.class.php:
```diff
@@ -15,74 +14,54 @@
+		$contactUrl = '';
+		return [
+			$this->getMessage( 'emailext-password-footer-1', $contactUrl )->parse(),
+			$this->getMessage( 'emailext-password-footer-2' )->escaped(),
```
I can't find those messages defined anywhere.

kvas-damian replied:
Oh, I know. This method is not used anywhere. Can we remove it?

garthwebb replied:
Ah good catch.

jsutterfield commented:
LGTM

Commit:
SOC-735 Removed unused function

Commit:
SOC-735 Added some docs

Commit:
SOC-735 Split out known argument from variable $params array

Commit:
Merge remote-tracking branch 'origin/dev' into SOC-735

Commit:
Merge branch 'email-sprint-15' into SOC-735

Commit:
SOC-735 Allow user objects to be passwed as parameters

Commit:
Merge branch 'email-sprint-15' into SOC-735
